virtual_rainforest/core/data.py

```diff
@@ -0,0 +1,686 @@
+
+Adding data from a file
+-----------------------
+
+The general solution for programmatically adding data from a file is to:
+
+* manually open a data file using the appropriate reader packages for the format,
+* coerce the data into a properly structured :class:`~xarray.DataArray` object, and then
+* use the :class:`~virtual_rainforest.core.data.Data.load_dataarray` method.
+
+However, the :meth:`~virtual_rainforest.core.data.Data.load_from_file` method
+automatically loads data from known formats defined in the
+:attr:`~virtual_rainforest.core.data.FILE_FORMAT_REGISTRY`.
+
+The FILE_FORMAT_REGISTRY
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The :attr:`~virtual_rainforest.core.data.FILE_FORMAT_REGISTRY` is used to register the
+set of known file formats for use in
+:meth:`~virtual_rainforest.core.data.Data.load_from_file`. This registry is extendable,
+so that new functions that implement the approach above for a given file format can be
+added to those supported by :meth:`~virtual_rainforest.core.data.Data.load_from_file`.
+This is done using the :func:`~virtual_rainforest.core.data.register_file_format_loader`
+decorator, which needs to specify the file formats supported (as a tuple of file
+suffixes) and then decorate a function that returns a :class:`~xarray.DataArray`
+suitable for use in :meth:`~virtual_rainforest.core.data.Data.load_dataarray`. For
+example:
+
+.. code-block:: python
+
+    @register_file_format_loader(('.tif', '.tiff'))
+    def new_function_to_load_tif_data(...):
+
+Using a data configuration
+--------------------------
+
+A :class:`~virtual_rainforest.core.data.Data` instance can also be populated using the
+:meth:`~virtual_rainforest.core.data.Data.load_from_config` method. This is expecting to
+take a properly validated configuration dictionary, loaded from a TOML file that
+specifies data source files:
+
+.. code-block:: toml
+
+    [[core.data.variable]]
+    name="precipitation"
+    file="/path/to/file.nc"
+    file_var="precip"
+    [[core.data.variable]]
+    name="temperature"
+    file="/path/to/file.nc"
+    file_var="temp"
+    [[core.data.variable]]
+    name="elevation"
+    file="/path/to/file.csv"
+    file_var="elev"
+
+Note that the properties for each variable in the configuration file are just the
+arguments for :meth:`~virtual_rainforest.core.data.Data.load_from_file`. The `replace`
+argument is not supported: data configurations should not contain multiple definitions
+of the same variable and so will fail with repeated variable names.
+"""
+
+from collections import UserDict
+from pathlib import Path
+from typing import Any, Callable, Optional
+
+import numpy as np
+from xarray import DataArray, Dataset, load_dataset
+
+from virtual_rainforest.core.config import ConfigurationError
+from virtual_rainforest.core.grid import GRID_REGISTRY, Grid
+from virtual_rainforest.core.logger import LOGGER, log_and_raise
+
+FILE_FORMAT_REGISTRY: dict[str, Callable] = {}
```

Comment:
On top of what Alex says, the independent readers should live in its own `readers.py` file or something like that, not here.

Reply:
Splitting `readers.py` definitely makes sense. It also would make some sense to split out the validators and the validator registry. But... then:

* `data.py` would need `from virtual_rainforest.core.validators import AXIS_VALIDATORS` 
* `validators.py` would need `from virtual_rainforest.core.data import Data`

And those are circular imports - is there a good way to keep small files with separate functionality and avoid this?

Reply:
The traditional approach is to import one of them - in this case probably the `AXIS_VALIDATORS` - locally within the function/method using it, rather than globally.